source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,875 @@
+difference is that change streams are resumable and drivers automatically
+perform resume attempts when they encounter transient errors. This allows
+change streams to be resilient to timeouts. If ``timeoutMS`` expires during a
+next call, drivers can’t auto-resume, but they can make sure the change
+stream is not invalidated so the user can call next again. In this case, the
+subsequent call would perform the resume without doing a ``getMore`` first.
+
+withTransaction communicates ``timeoutMS`` via ClientSession
+--------------------------------------------------------
+
+Because the ``withTransaction`` API doesn’t allow drivers to plumb down the
+remaining timeout into the user-provided callback, this spec requires the
+remaining timeout to be stored on the ClientSession. Operations in the
+callback that run under that ClientSession can then extract the timeout from
+the session and apply it. To avoid confusing validation semantics, operations
+error if there is a timeout on the session but also an overridden timeout for
+the operation. It’s possible that the ability to communicate timeouts for a
+block of operations via a ClientSession is useful as a general purpose API,
+but we’ve decided to make it private until there are other known use cases.
+
+withTransaction refreshes the timeout for abortTransaction
+----------------------------------------------------------
+
+If the user-provided callback to ``withTransaction`` times out, it could
+leave a transaction running on the server. It’s imperative that drivers make
+an effort to abort the open transaction because failing to do so could result
+in the collections and databases affected by the transaction being locked for
+a long period of time, which could cause applications to stall. Because
+``timeoutMS`` has expired before drivers attempt to abort the transaction, we
+require drivers to refresh it and apply the original value to the execution
+of the ``abortTransaction`` operation. This can cause the entire
+``withTransaction`` call to take up to ``2*timeoutMS``, but it was decided
+that this risk is worthwhile given the importance of transaction cleanup.
+
+GridFS streams behavior
+-----------------------
+
+Streams created by GridFS API operations (e.g. by ``open_upload_stream`` and
+``open_download_stream``) present a challenge for this specification. These
+types of streams execute multiple operations, but there can be artificial
+gaps between operations if the application does not invoke the stream
+functions for long periods of time. Generally, we expect users to upload or
+download an entire file as quickly as possible, so we decided to have
+``timeoutMS`` cap the lifetime of the created stream. The other option was to
+apply the entire ``timeoutMS`` value to each operation executed by the
+stream, but streams perform many hidden operations, so this approach could
+cause an upload/download to take much longer than expected.
```

Comment:
What is the concern with applying `timeoutMS` across database operations executed within the lifetime of a stream? Are you concerned about overhead of granular timeout calculations?

Reply:
My concern was that users are likely not aware of all of the operations that a single stream will do. For example, uploading a single file can actually result in multiple index and collection operations. If we refreshed `timeoutMS` for each one, a user might expect the upload to either complete or time out after some amount of time, but it would actually take much longer due to all of the hidden operations.

Reply:
> a user might expect the upload to either complete or time out after some amount of time, but it would actually take much longer due to all of the hidden operations.

My concern isn't that operations might take _longer_ than expected. On the contrary, I think a user might be surprised that `timeoutMS` is reached _sooner_ than expected.

To clarify, I wasn't proposing that we use the original `timeoutMS` value for each internal operation such that we'd risk spending some multiple of `timeoutMS` across several slow, internal operations. I assume that's what you mean by "refreshing".

My question here went along with https://github.com/mongodb/specifications/pull/881/files#r561776825, where I assumed that we'd exclude any time spent doing IO on the user's stream from counting against `timeoutMS`. And I certainly didn't expect it to count more generally for the amount of time the GridFS object exists in memory.

Reply:
I'll lump this in with the other GridFS question in an offline conversation as they seem related.